### PR TITLE
Create Report for the given file ID

### PR DESCRIPTION
**Pull request by kirankumbhar**
| Title                          | Value                                                                                     |
| ------------------------------ | ----------------------------------------------------------------------------------------- |
| **Type**                       | Feature / Bugfix / Refactor / Documentation                                               |
| **Ticket/Issue**               | PD 721                                      |
| **Migrations**                 | No         |
| **Migration&nbsp;Scripts**     | No                                                |
| **ENV&nbsp;vars&nbsp;change**  | No |
| **Frontend**                   | No                                                            |
| **Local&nbsp;testing**         | Done                                                                            |
| **Staging&nbsp;testing**       |Skip                                              |
| **On&nbsp;premise&nbsp;notes** | NA                               |
| **Documentation**              | None                       |
| **Release&nbsp;notes**         | None                               |
| **Version&nbsp;upgrade**       | Major                                                                 |

---

### Changelog

1. Added create report command

  `appknox reports create fileID`

### CODE REVIEW

File: helper/reports.go [hunk 1]
```diff
@@ -80,3 +80,21 @@ func ProcessDownloadReportCSV(reportID int, outputFilePath string) error {
...
+func ProcessCreateReport(fileID int) (reportID int, err error) {
+	ctx := context.Background()
+	client := getClient()
+	reportID, err = client.Reports.CreateReport(ctx, fileID)
+	if err != nil && err.Error() == "400 Bad Request" {
+		results, err := client.Reports.List(ctx, fileID)
+		if err != nil {
+			return 0, err
+		}
+		return results[0].ID, nil
```
Comment (rajan262) [inline helper/reports.go]: Getting non-empty report list can't be guaranteed. The scenario where report list is empty should ideally be handled here.

File: appknox/reports.go [hunk 2]
```diff
@@ -137,3 +137,17 @@ func (s *ReportsService) WriteReportDataToFile(respBody io.ReadCloser, outputFil
...
+	request, err := s.client.NewRequest("POST", url, nil)
+	var reportResult ReportResult
+	resp, err := s.client.Do(ctx, request, &reportResult)
+	if resp != nil && resp.StatusCode == 400 {
+		return 0, errors.New(resp.Status)
+	}
+	if err != nil {
+		return 0, err
+	}
+	return reportResult.ID, nil
```
Comment (rajan262) [inline appknox/reports.go]: I think CreateReport should return the whole ReportResult instance. It should be up to helper or cmd to decide what to take out of that instance.

File: appknox/appknox.go [hunk 3]
```diff
@@ -230,6 +230,7 @@ func (c *Client) Do(ctx context.Context, req *http.Request, v interface{}) (*Res
 	}
 	if v != nil {
 		if w, ok := v.(io.Writer); ok {
+
```
Comment (rajan262) [inline appknox/appknox.go]: This diff is unnecessary. Remove it.

File: cmd/reports.go [hunk 4]
```diff
@@ -36,6 +37,32 @@ var reportsListCmd = &cobra.Command{
 		}
 	},
 }
+var reportsCreateCmd = &cobra.Command{
+	Use:   "create",
+	Short: "Create Report for the given file ID.",
+	Long: `Triggers report generation if report doesn't exists or not updated
```
Comment (cosmosgenius) [inline cmd/reports.go]: Lets not handle 400. As ideally 400 should not occur let the backend handle it if required. For the now the message should be shown to the user

**Review by cosmosgenius — CHANGES_REQUESTED** [review]

File: helper/reports.go (hunk 1 above)
Comment (kirankumbhar) [reply helper/reports.go]: This is no longer valid As we are not going with fallback option of getting latest report ID

File: appknox/reports.go (hunk 2 above)
Comment (kirankumbhar) [reply appknox/reports.go]: Done 👍🏼

File: cmd/reports.go (hunk 4 above)
Comment (kirankumbhar) [reply cmd/reports.go]: Done 👍🏼

File: appknox/appknox.go (hunk 3 above)
Comment (kirankumbhar) [reply appknox/appknox.go]: Done 👍🏼

**Review by cosmosgenius — APPROVED** [review]

**Pull request merged**